docs/development/mono-repository.md

```diff
@@ -0,0 +1,120 @@
+
+## Recommended Packages
+
+To enable a mono-repository, we recommend leveraging [yarn workspaces][yarn-workspaces] or [npm workspaces][npm-workspaces].
+
+### Setup with Yarn Example
+
+#### Modify package.json in the workspace root
+
+```js
+{
+  "private": true,
+  // this is an array containing glob paths to each workspace
+  "workspaces": ["workspace-a", "packages/*"]
+}
+```
+
+#### Create subfolders and package.json for each package in the monorepo
+
+```js
+// /workspace-a/package.json
+{
+  "name": "workspace-a",
+  "version": "1.0.0",
+  "dependencies": {
+    "foo": "0.0.1"
+  }
+}
+
+// /packages/foo/package.json
+{
+  "name": "foo",
+  "version": "1.0.0"
+}
+```
+
+#### Run `yarn install` in the workspace root
+
+
+Given the above, we should expect the following folder structure
+```
+/package.json
+/yarn.lock
+/workspace-a/package.json
+/packages/foo/package.json
+/node_modules
+/node_modules/workspace-a -> /workspace-a
+```
+
+## Guidance
+
+There are a handful of benefits and challenges to monorepos.
+
+With a monorepo, it is recommended to leverage a common set of tooling for [linting][linting] and [testing][testing] across all packages in the repository. See our [code-consistency][code-consistency] section for further guidance. Having a common set of tooling is a benefit for monorepos since each package managed in the repo does not need its set of tooling or versioning. On the other hand, teams can also control scripts individually per artifact or package if necessary. See `Advanced Tips / Tricks` section for more details.
+
+While there are a benefits, there are challenges to overcome.
+
+* Detecting what to build. Build times can be long due to overbuilding the entire repository
+* Long build times. Because of having to build the entire repository, this will cause build times to increase because tooling no longer is just building one artifact/package, but many different ones.
+* Not just JavaScript. A monorepo may not just be JavaScript, and can impact organization, tooling, and how to approach building different artifacts or packages in different languages.
+* Git repositories can grow larger in size.  Since we are building and maintaing different packages in a singular repository, we can exponentially grow the size of the git repository.
+* Commit messages can be harder to interpret due to multiple artifacts, packages, or teams.
+
+
+### Building
+
+When building npm packages (or any type of package) (the build) in the monorepo, we recommend:
+
+* [bazel][bazel]
```

Comment:
This recommendation is puzzling. javascript ecosystem does not seem to be a priority in bazel documentation. For example, their [tutorials](https://bazel.build/start/getting-started) do not list js/ts or web

Reply:
@jimlindeman could you provide any feedback for your usage of bazel?

Reply:
We had used Bazel primarily because it 
1) provided a generic mechanism to support multiple programming languages (for example typescript for the primary app language and golang for the operator/CRN definition to configure/deploy the app)
2) It let us easily hierarchically define what components had to be compiled and tested based on which files or dependent libraries changed

Reply:
Example BUILD file for the logging package in our monorepo:
```
load("@npm_bazel_typescript//:index.bzl", "ts_library")
load("//hack:typescript.bzl", "unit_test")

package(default_visibility = ["//visibility:public"])

deps = [
    "//ts/packages/constants",
    "//ts/packages/utils",
    "@npm//@types/cls-hooked",
    "@npm//@types/convict",
    "@npm//@types/morgan",
    "@npm//@types/node",
    "@npm//chalk",
    "@npm//cls-hooked",
    "@npm//convict",
    "@npm//logform",
    "@npm//morgan",
    "@npm//winston",
    "@npm//winston-transport",
]

ts_library(
    name = "logger",
    srcs = glob(
        ["**/*.ts"],
        [
            "tests/**/*.ts",
            "**/*.test.ts",
            "**/*.spec.ts",
            "node_modules/**/*.ts",
        ],
    ),
    runtime = "nodejs",
    deps = deps,
)

test_deps = [
    "@npm//@types/sinon",
    "@npm//sinon",
]

unit_test(deps + test_deps)
```
So in this example the top two dependencies were other packages (essentially folders) in the same monorepo, so that if they changes this package would also need to get rebuilt/tested.  The other dependencies were to track if those external libraries changed with fix-version changes.

Reply:
would it make sense to recommend bazel specifically for monorepos that make use of multiple non-javascript backend languages?

Maybe I'm being needlessly puritanical, but a substantial tool like bazel may be inappropriate for all-node monorepos, so making it a default 'blanket' recommendation could end up adding tooling burden, even if it's genuinely appropriate for more complex projects like the golang project you described.

Reply:
I would love to mention `nx` and `rush` here, but since the existing recommendations are our experience, I would say we cannot recommend it. Maybe @mhdawson do you have recommendations for how to present these "honorable mentions"?

Reply:
They look promising from a pure Node.js perspective, we wouldn't use them simply because we have to support multiple languages though.